packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,335 @@
+import { MultisigAddressType, Network, getAddressType } from "@caravan/bitcoin";
+
+/*
+The methodology for calculating a privacy score (p_score) for Bitcoin transactions based 
+on the number of inputs and outputs is the primary point to define wallet health for privacy. 
+The score is further influenced by several factors such as address reuse, 
+address types and UTXO set fingerprints etc.
+
+More on the algorithms for scoring privacy: <TODO : Add Link to Blog Post>
+*/
+
+// A normalizing quantity that increases the score by a certain factor in cases of self-payment.
+// More about deniability : https://www.truthcoin.info/blog/deniability/
+const DENIABILITY_FACTOR = 1.5;
+
+/*
+The p_score is calculated by evaluating the likelihood of self-payments, the involvement of 
+change outputs and the type of transaction based on number of inputs and outputs.
+
+We have 5 categories of transaction type
+- Sweep Spend
+- Simple Spend
+- UTXO Fragmentation (any transaction with more than the standard 2 outputs)
+- Consolidation
+- CoinJoin
+*/
+enum SpendType {
+  SweepSpend = "SweepSpend",
+  SimpleSpend = "SimpleSpend",
+  UTXOFragmentation = "UTXOFragmentation",
+  Consolidation = "Consolidation",
+  MixingOrCoinJoin = "MixingOrCoinJoin",
+}
+
+/*
+The deterministic scores or their formula for each spend type are as follows
+*/
+function spendTypeScores(
+  spendType: SpendType,
+  numberOfInputs: number,
+  numberOfOutputs: number,
+): number {
+  switch (spendType) {
+    case SpendType.SweepSpend:
+      return 1 / 2;
+    case SpendType.SimpleSpend:
+      return 4 / 9;
+    case SpendType.UTXOFragmentation:
+      return 2 / 3 - 1 / numberOfOutputs;
+    case SpendType.Consolidation:
+      return 1 / numberOfInputs;
+    case SpendType.MixingOrCoinJoin:
+      let x = Math.pow(numberOfOutputs, 2) / numberOfInputs;
+      return ((2 / 3) * x) / (1 + x);
+    default:
+      throw new Error("Invalid spend type");
+  }
+}
+
+function determineSpendType(inputs: number, outputs: number): SpendType {
+  if (inputs === 1) {
+    if (outputs === 1) return SpendType.SweepSpend;
+    if (outputs === 2) return SpendType.SimpleSpend;
+    return SpendType.UTXOFragmentation;
+  } else {
+    if (outputs === 1) return SpendType.Consolidation;
+    return SpendType.MixingOrCoinJoin;
+  }
+}
+
+/*
+The transaction topology refers to the type of transaction based on 
+number of inputs and outputs.
+
+Expected Range : [0, 0.75]
+-> Very Poor : [0, 0.15]
+-> Poor : (0.15, 0.3]
+-> Moderate : (0.3, 0.45]
+-> Good : (0.45, 0.6]
+-> Very Good : (0.6, 0.75] 
+*/
+export async function privacyScoreByTxTopology(
+  transactions: Transaction[],
+  client: BlockchainClient,
+): Promise<number> {
+  let privacyScore = 0;
+  for (let tx of transactions) {
+    let topologyScore = await scoreForTxTopology(tx, client);
+    privacyScore += topologyScore;
+  }
+  return privacyScore / transactions.length;
+}
+
+export async function scoreForTxTopology(
```

Comment:
Where are we finding the median topology score ? Let me know if you think we are finding the median score anywhere because we are just focussing on the mean score from tx topology pov. 

I can explain the thought process behind mean score here.

So the topology score for a single tx is designed such that, the topology score itself speaks of how good privacy wise the wallet has been for that single particular transaction.

Taking a mean of all the past transactions in wallet history will give us an estimate of on an avg how the given wallet is performing on the privacy.

An example to explain the above thought.
 A wallet makes 5 transactions. Let the score be ( 0<=score<=1 ). 0 is bad privacy and 1 is best privacy.
Tx 1 - 0.9
Tx 2 - 0.9
Tx 3 - 0.8
Tx 4 - 0.1
Tx 5 - 0.8
--------------
Mean score over all txs is 0.7 
If the wallet owner is not okay with such a score, They can choose to perform actions that increases privacy. Now the number 0.7 is good or bad for privacy depends on the consumer of @caravan/health and they get to decide their own math. 

If they want they can take a median value as well.  Since we are also exposing it for a single transaction, doing the same calculations with median is possible on the consumer side.

Reply:
> Where are we finding the median topology score ? Let me know if you think we are finding the median score anywhere because we are just focussing on the mean score from tx topology pov.

No, I don't think we are, i was just kind of curious if you thought it was meaningful at all.

Reply:
It's doable at consumer level too if not from health.